src/Workspaces/Remote/Core/Serialization/ImmutableCollectionMessagePackResolver.cs

```diff
@@ -0,0 +1,394 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Reflection;
+using MessagePack;
+using MessagePack.Formatters;
+
+// TODO: Copied from https://github.com/neuecc/MessagePack-CSharp/blob/master/src/MessagePack.ImmutableCollection/Formatters.cs.
+// Replace with an implementation shipping with VS:
+// https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1198374
+// https://github.com/neuecc/MessagePack-CSharp/issues/606
+//
+// The following code includes fix for: https://github.com/neuecc/MessagePack-CSharp/issues/1033. Make sure the fix is included in the replacement.
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal sealed class ImmutableCollectionMessagePackResolver : IFormatterResolver
+    {
+        public static readonly ImmutableCollectionMessagePackResolver Instance = new ImmutableCollectionMessagePackResolver();
+
+        private ImmutableCollectionMessagePackResolver()
+        {
+        }
+
+        public IMessagePackFormatter<T> GetFormatter<T>()
+            => FormatterCache<T>.Formatter;
+
+        private static class FormatterCache<T>
+        {
+            internal static readonly IMessagePackFormatter<T> Formatter;
+
+            static FormatterCache()
+            {
+                Formatter = (IMessagePackFormatter<T>)GetFormatter(typeof(T));
+            }
+        }
+
+        private static readonly Dictionary<Type, Type> s_formatterMap = new Dictionary<Type, Type>()
+        {
+            { typeof(ImmutableArray<>), typeof(ImmutableArrayFormatter<>) },
+            { typeof(ImmutableList<>), typeof(ImmutableListFormatter<>) },
+            { typeof(ImmutableDictionary<,>), typeof(ImmutableDictionaryFormatter<,>) },
+            { typeof(ImmutableHashSet<>), typeof(ImmutableHashSetFormatter<>) },
+            { typeof(ImmutableSortedDictionary<,>), typeof(ImmutableSortedDictionaryFormatter<,>) },
+            { typeof(ImmutableSortedSet<>), typeof(ImmutableSortedSetFormatter<>) },
+            { typeof(ImmutableQueue<>), typeof(ImmutableQueueFormatter<>) },
+            { typeof(ImmutableStack<>), typeof(ImmutableStackFormatter<>) },
+            { typeof(IImmutableList<>), typeof(InterfaceImmutableListFormatter<>) },
+            { typeof(IImmutableDictionary<,>), typeof(InterfaceImmutableDictionaryFormatter<,>) },
+            { typeof(IImmutableQueue<>), typeof(InterfaceImmutableQueueFormatter<>) },
+            { typeof(IImmutableSet<>), typeof(InterfaceImmutableSetFormatter<>) },
+            { typeof(IImmutableStack<>), typeof(InterfaceImmutableStackFormatter<>) },
+        };
+
+        internal static object GetFormatter(Type t)
```

Comment:
this is def worth doc'ing the impl.  also, does this need to be internal?

Reply:
See the comment at the beginning of the file.